plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright (C) 2021 Jarvis Jiang <jarvis.w.jiang@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <sys/stat.h>
+#include <stdio.h>
+
+#include <fwupdplugin.h>
+#include "fu-mbim-qdu-updater.h"
+#include "fu-mm-utils.h"
+
+#if MBIM_CHECK_VERSION(1,25,3)
+#define FU_MBIM_QDU_MAX_OPEN_ATTEMPTS 8
+
+struct _FuMbimQduUpdater {
+	GObject		 parent_instance;
+	gchar		*mbim_port;
+	MbimDevice	*mbim_device;
+};
+
+G_DEFINE_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+	guint		 open_attempts;
+} OpenContext;
+
+static void fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
```

Comment:
I think you should have an assertion check here to ensure that `open_attempts` is > 0.  If you end up somehow back into this function with 0 it will wrap and can get stuck.